### ISSUE TITLE
[gcp] New plugin for Google Cloud Platform

### ISSUE THREAD

**Issue by m-strzelczyk** [sosreport/sos#2699]
Hi!

I'm planning to write a new plugin to better support Virtual Machines running on Google Cloud Platform. Similar to what is already implemented for [Azure](https://github.com/sosreport/sos/blob/main/sos/report/plugins/azure.py).

I plan to capture following info:
 * All information that can be extracted from [Metadata server](https://cloud.google.com/compute/docs/metadata/overview).
 * Get output of `gcloud auth list` - if it's available
 * Grab journal logs for any Google-related services 

When would it run?
Only on VMs running on GCP - meaning a `curl` to `http://metadata.google.internal/computeMetadata/v1/` is successful.

My questions:
 * Is `gcp` a good name for such plugin, or would `google-cloud` be better?
 * Should all this data be collected inside one plugin, or is it better to divide those into for example: `gcp-metadata` and `gcp-services`?

**Comment by pmoravec** [comment]
That plugin can have sense.

> When would it run?
> Only on VMs running on GCP - meaning a `curl` to `http://metadata.google.internal/computeMetadata/v1/` is successful.

You mean the plugin enabledness condition would be the curl request? Assume a system (not gCloud VM) with DNS issues where such curl would hang - then sos report would hang even during startup phase. Can't be a test like "is `gcloud` executable?" work well?

> My questions:
> 
>     * Is `gcp` a good name for such plugin, or would `google-cloud` be better?

No strong opinion here. On one side, "gcp" is quite concise such that one must know the acronym to understand it. On the other side, the acronym is well(?) known.

>     * Should all this data be collected inside one plugin, or is it better to divide those into for example: `gcp-metadata` and `gcp-services`?

If the data will be collected everytime altogether (under the same enablement conditions), I would stick with one plugin.

**Comment by TurboTurtle** [comment]
> > When would it run?
> > Only on VMs running on GCP - meaning a `curl` to `http://metadata.google.internal/computeMetadata/v1/` is successful.
> 
> You mean the plugin enabledness condition would be the curl request? Assume a system (not gCloud VM) with DNS issues where such curl would hang - then sos report would hang even during startup phase. Can't be a test like "is `gcloud` executable?" work well?
> 

Since this is for VMs in GCP, `gcloud` presence isn't a great indicator. You don't have to have `gcloud` installed on GCP VMs, and it can be installed on systems outside GCP to leverage GCP resources.

A `curl` while not ideal, could use a short timeout to avoiding hanging our setup phase.

Alternatively, is there some intrinsic attribute on the VM that could be used? E.G. something GCP specific from something like `dmesg` or `dmidecode`?

> > My questions:
> > ```
> > * Is `gcp` a good name for such plugin, or would `google-cloud` be better?
> > ```
> 
> No strong opinion here. On one side, "gcp" is quite concise such that one must know the acronym to understand it. On the other side, the acronym is well(?) known.
> 
> > ```
> > * Should all this data be collected inside one plugin, or is it better to divide those into for example: `gcp-metadata` and `gcp-services`?
> > ```
> 
> If the data will be collected everytime altogether (under the same enablement conditions), I would stick with one plugin.

Agree with single plugin. `gcp` seems fine for me.

**Comment by m-strzelczyk** [comment]
I agree that checking for `gcloud` being available is not a good indicator - developers install gcloud on their local machines to control cloud instances etc. And `gcloud` is definitely not guaranteed to be installed on VMs.

I have checked `dmesg` of an instance, and I like what I'm seeing. There is this line for example:
```
[    0.000000] DMI: Google Google Compute Engine/Google Compute Engine, BIOS Google 01/01/2011
```
Should be relatively easy to check.

As for using `dmidecode` - can we count on it being available? Is is part of every Linux distribution?

If `dmidecode` is really common, then it also has good indicators I guess:

```
~$ sudo dmidecode
# dmidecode 3.2
Getting SMBIOS data from sysfs.
SMBIOS 2.4 present.
14 structures occupying 546 bytes.
Table at 0xBE9CC000.
Handle 0x0000, DMI type 0, 24 bytes
BIOS Information
        Vendor: Google
        Version: Google
        Release Date: 01/01/2011
        Address: 0xE8000
        Runtime Size: 96 kB
        ROM Size: 64 kB
        Characteristics:
                BIOS characteristics not supported
                Targeted content distribution is supported
        BIOS Revision: 1.0
Handle 0x0097, DMI type 1, 27 bytes
System Information
        Manufacturer: Google
        Product Name: Google Compute Engine
        Version: Not Specified
        Serial Number: GoogleCloud-98E621D0FDC0576BB1A81BDCC7E66F61
        UUID: 98e621d0-fdc0-576b-b1a8-1bdcc7e66f61
        Wake-up Type: Power Switch
        SKU Number: Not Specified
        Family: Not Specified
Handle 0x0098, DMI type 2, 15 bytes
Base Board Information
        Manufacturer: Google
        Product Name: Google Compute Engine
        Version: Not Specified
        Serial Number: Board-GoogleCloud-98E621D0FDC0576BB1A81BDCC7E66F61
        Asset Tag: 98E621D0-FDC0-576B-B1A8-1BDCC7E66F61
        Features:
                Board is a hosting board
        Location In Chassis: Not Specified
        Chassis Handle: 0x0099
        Type: Motherboard
        Contained Object Handles: 0
```

**Comment by TurboTurtle** [comment]
> I have checked `dmesg` of an instance, and I like what I'm seeing. There is this line for example:
> 
> ```
> [    0.000000] DMI: Google Google Compute Engine/Google Compute Engine, BIOS Google 01/01/2011
> ```
> 
> Should be relatively easy to check.
> 
> As for using `dmidecode` - can we count on it being available? Is is part of every Linux distribution?
> 

It's part of every Linux distro we currently support. Beyond that, I would honestly be beside myself if there was a distro out there not shipping either/both of these.
